internal/resources/iampolicy/scope.go

```diff
@@ -0,0 +1,197 @@
+	scopedFullname struct {
+		scope                scope
+		fullnameOrganization *organizationmodel.VmwareTanzuManageV1alpha1OrganizationFullName
+		fullnameClusterGroup *clustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupFullName
+		fullnameCluster      *clustermodel.VmwareTanzuManageV1alpha1ClusterFullName
+		fullnameWorkspace    *workspacemodel.VmwareTanzuManageV1alpha1WorkspaceFullName
+		fullnameNamespace    *namespacemodel.VmwareTanzuManageV1alpha1ClusterNamespaceFullName
+	}
+)
+
+func constructScope(d *schema.ResourceData) (scopedFullnameData *scopedFullname) {
+	value, ok := d.GetOk(scopeKey)
+	if !ok {
+		return scopedFullnameData
+	}
+
+	data, _ := value.([]interface{})
+
+	if len(data) == 0 || data[0] == nil {
+		return scopedFullnameData
+	}
+
+	scopeData := data[0].(map[string]interface{})
+
+	if v, ok := scopeData[organizationKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:                organizationScope,
+				fullnameOrganization: constructOrganizationFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[clusterGroupKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:                clusterGroupScope,
+				fullnameClusterGroup: constructClusterGroupFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[clusterKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:           clusterScope,
+				fullnameCluster: constructClusterFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[workspaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:             workspaceScope,
+				fullnameWorkspace: constructWorkspaceFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[namespaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:             namespaceScope,
+				fullnameNamespace: constructNamespaceFullname(v1),
+			}
+		}
+	}
+
+	return scopedFullnameData
+}
+
+func flattenScope(scopedFullname *scopedFullname) (data []interface{}) {
+	if scopedFullname == nil {
+		return data
+	}
+
+	flattenScopeData := make(map[string]interface{})
+
+	switch scopedFullname.scope {
+	case organizationScope:
+		flattenScopeData[organizationKey] = flattenOrganizationFullname(scopedFullname.fullnameOrganization)
+	case clusterGroupScope:
+		flattenScopeData[clusterGroupKey] = flattenClusterGroupFullname(scopedFullname.fullnameClusterGroup)
+	case clusterScope:
+		flattenScopeData[clusterKey] = flattenClusterFullname(scopedFullname.fullnameCluster)
+	case workspaceScope:
+		flattenScopeData[workspaceKey] = flattenWorkspaceFullname(scopedFullname.fullnameWorkspace)
+	case namespaceScope:
+		flattenScopeData[namespaceKey] = flattenNamespaceFullname(scopedFullname.fullnameNamespace)
```

Comment:
Please import and use the flatten full name functions as the base resources are already implemented.